rfhub2/cli/test_cases/test_cases_extractor.py

```diff
@@ -0,0 +1,121 @@
+from pathlib import Path
+from robot.running.model import TestSuite
+from robot.testdoc import TestSuiteFactory
+from typing import List, Tuple
+
+
+@dataclass
+class Keyword:
+    name: str
+
+
+@dataclass
+class Test:
+    doc: str
+    id: str
+    longname: str
+    name: str
+    parent: str
+    source: str
+    tags: List[str]
+    template: str
+    timeout: str
+    keywords: List[Keyword]
+    setup: List[Keyword]
+    teardown: List[Keyword]
+
+
+@dataclass
+class Suite:
+    doc: str
+    id: str
+    longname: str
+    name: str
+    parent: str
+    source: str
+    test_count: int
+    tests: List[Test]
+    setup: List[Keyword]
+    teardown: List[Keyword]
+
+
+class TestCasesExtractor:
+    def __init__(self, paths: Tuple[Path, ...]) -> None:
+        self.paths = paths
+
+    def create_testdoc_from_paths(self):
+        """
+        Creates TestDoc objects based on robot.testdoc.TestSuiteFactory method.
+        """
+        return [suite for path in self.paths for suite in self.create_testdoc_from_path(path)]
+
+    def create_testdoc_from_path(self, path: Path):
+        """
+        Creates TestDoc object based on robot.testdoc.TestSuiteFactory method.
+        """
+        testdoc = TestSuiteFactory(str(path))
+        return [self.serialize_testdoc(testdoc)] + self._traverse_suites(testdoc)
+
+    def _traverse_suites(self, testdoc: TestSuite) -> List[Suite]:
+        """
+        Traverses all suites in top level suites and returns suites list.
+        """
+        suites = []
+        for suite in testdoc.suites:
+            suites.append(self.serialize_testdoc(suite))
+            sub_suites = self._traverse_suites(suite)
+            suites += sub_suites
+        return suites
+
+    def serialize_testdoc(self, testdoc: TestSuite) -> Suite:
+        """
+        Serializes testdoc object into Suite dataclass object.
+        """
+        return Suite(doc=testdoc.doc, id=testdoc.id, longname=testdoc.longname, name=testdoc.name,
+                     parent=testdoc.parent._name if hasattr(testdoc.parent, '_name') else None,
+                     source=testdoc.source, test_count=testdoc.test_count, tests=self.get_tests(testdoc),
+                     setup=self.get_setup_keywords(testdoc), teardown=self.get_teardown_keywords(testdoc))
+
+    def get_tests(self, testdoc: TestSuite) -> List[Test]:
+        """
+        Returns list of tests.
+        """
+        return [Test(doc=test.doc, id=test.id, longname=test.longname, name=test.name, parent=test.parent.name,
+                     source=test.source if hasattr(test, 'source') else None,
+                     tags=list(test.tags), template=test.template, timeout=test.timeout,
+                     keywords=self.get_normal_keywords(test), setup=self.get_setup_keywords(test),
+                     teardown=self.get_teardown_keywords(test)) for test in testdoc.tests]
+
+    def get_setup_keywords(self, testdoc: TestSuite) -> List[Keyword]:

```

Comment:
These 3 methods could be turned into single method that accepts Union[List[Keyword], Keyword, None]

Reply:
Yeah, I just thought of using has.attr, that would be used with method parameter, like normal, teardown, setup. I will add some changes tomorrow

Reply:
I did some changes for this one, but I don't think it looks good. I believe I didn;t get what You had in mind. Anyway, I merged setup/teardown keyword into one method with kw_type parameter, and simplified get_normal_keywords method into single line list comprehension.